tools/QuietBackgroundProcesses/DevHome.QuietBackgroundProcesses.Server/QuietBackgroundProcessesSessionManager.cpp

```diff
@@ -0,0 +1,107 @@
+
+#include <QuietBackgroundProcesses.h>
+
+#include "DevHome.QuietBackgroundProcesses.QuietBackgroundProcessesSession.h"
+#include "DevHome.QuietBackgroundProcesses.QuietBackgroundProcessesSessionManager.h"
+
+#define STATUS_ACCESS_DENIED ((NTSTATUS)0xC0000022L)
+#define STATUS_OBJECT_NAME_NOT_FOUND ((NTSTATUS)0xC0000034L)
+
+static bool isFeatureSupported() noexcept
+{
+    HRESULT hr = DisableQuietBackgroundProcesses();
+    if (hr == HRESULT_FROM_NT(STATUS_OBJECT_NAME_NOT_FOUND))
+    {
+        return false;
+    }
+    else if (hr == HRESULT_FROM_NT(STATUS_ACCESS_DENIED))
+    {
+        return true;
+    }
+    LOG_HR(hr);
+    return false;
+}
+
+namespace ABI::DevHome::QuietBackgroundProcesses
+{
+    class QuietBackgroundProcessesSessionManager :
+        public Microsoft::WRL::RuntimeClass<
+            Microsoft::WRL::RuntimeClassFlags<Microsoft::WRL::WinRt>,
+            IQuietBackgroundProcessesSessionManager,
+            Microsoft::WRL::FtmBase>
+    {
+        InspectableClass(RuntimeClass_DevHome_QuietBackgroundProcesses_QuietBackgroundProcessesSessionManager, BaseTrust);
+
+    public:
+        STDMETHODIMP RuntimeClassInitialize() noexcept
+        {
+            return S_OK;
+        }
+    };
+
+    class QuietBackgroundProcessesSessionManagerStatics WrlFinal :
+        public Microsoft::WRL::AgileActivationFactory<
+            Microsoft::WRL::Implements<IQuietBackgroundProcessesSessionManagerStatics>>
+    {
+        InspectableClassStatic(RuntimeClass_DevHome_QuietBackgroundProcesses_QuietBackgroundProcessesSessionManager, BaseTrust);
+
+    public:
+        // IActivationFactory method
+        STDMETHODIMP ActivateInstance(_Outptr_result_nullonfailure_ IInspectable** ppvObject) noexcept
+        try
+        {
+            THROW_IF_FAILED(Microsoft::WRL::MakeAndInitialize<QuietBackgroundProcessesSessionManager>(ppvObject));
+            return S_OK;
+        }
+        CATCH_RETURN()
+
+        // IQuietBackgroundProcessesSessionManagerStatics
+        STDMETHODIMP IsFeatureSupported(_Out_ boolean* isSupported) noexcept override
+        {
+            *isSupported = isFeatureSupported();
+            return S_OK;
+        }
+
+        STDMETHODIMP GetSession(_Outptr_result_nullonfailure_ IQuietBackgroundProcessesSession** session) noexcept override
+        try
+        {
+            auto lock = std::scoped_lock(m_mutex);
+            *session = nullptr;
+
+            if (!m_sessionReference)
+            {
+                auto factory = wil::GetActivationFactory<IQuietBackgroundProcessesSessionStatics>(RuntimeClass_DevHome_QuietBackgroundProcesses_QuietBackgroundProcessesSession);
+                THROW_IF_FAILED(factory->GetSingleton(&m_sessionReference));
+            }
+            m_sessionReference.copy_to(session);
+            return S_OK;
+        }
+        CATCH_RETURN()
+
+        STDMETHODIMP TryGetSession(_COM_Outptr_result_maybenull_ IQuietBackgroundProcessesSession** session) noexcept override try
```

Comment:
>TryGetSession [](http://example.com/codeflow?start=21&length=13)

Note that the current semantics of TryGetSession are basically "somebody called GetSession at some point". It does not actually verify that the elevated session process is still around, since the QI under try_copy_to would be handled purely within the client-side proxy code.

If you want to force an interaction with the server to confirm it's still actually around, you could always add a Ping() method or something along those lines.

Reply:
I don't get it - if m_sessionReference is valid once, won't it keep the server alive?  In what circumstance could the server go away but m_sessionReference still have a refcount of 1?

Reply:
Server crashed or got externally killed or something? Not part of the golden path, but worth considering how that case is handled.

Reply:
Seems (mostly) futile - there's a permanent race anyway if you Ping() then the server goes down before you use it.  I'm treating the elevated server as super stable un-crashable because it's putting the system into quiet mode  - we do not want the machine stuck in that state indefinitely.  We also can't restart the server without re-issuing a UAC prompt.

(Read: I don't want to do this, but I will if you think it's the right thing to do.)

Reply:
So if the elevated sever crashes, are we just perma-wedged?

I guess that's the core of what I'm concerned about here: can we recover when unexpected terminations happen? I don't really care whether that recovery is driven from this manager process or from DevHome itself, but I don't think we want to end up in a state where the only way out of quiet mode is rebooting the computer.